CLI and lang updates assoc with tagging (#35668)
hashicorp/terraform: website/docs/language/terraform.mdx

---
page_title: Terraform block configuration reference
description: >-
  The `terraform` block allows you to configure Terraform behavior, including the Terraform version, backend, integration with HCP Terraform, and required providers.
---

# Terraform block reference

This topic provides reference information about the `terraform` block. The `terraform` block allows you to configure Terraform behavior, including the Terraform version, backend, integration with HCP Terraform, and required providers.

## Configuration model

The following list outlines attribute hierarchy, data types, and requirements in the `terraform` block. Click on an attribute for details.

- [`terraform`](#terraform)

>   - [`required_version`](#terraform-required_version): string | defaults to the latest version

Reviewer: I don't think this defaults-- omitting it creates no version requirement.

>   - [`required_version`](#terraform-required_version): string | defaults to the latest version

Author: ```suggestion
  - [`required_version`](#terraform-required_version): string 
```

  - [`required_providers`](#terraform-required_providers): map
  - [`provider_meta "<LABEL>"`](#terraform-provider_meta): map
  - [`backend "<BACKEND_TYPE>"`](#terraform-backend): map
  - [`cloud`](#terraform-cloud): map
    - [`organization`](#terraform-cloud-organization): string | required when connecting to HCP Terraform
    - [`workspaces`](#terraform-cloud-workspaces): map | required when connecting to HCP Terraform
      - [`tags`](#terraform-cloud-workspaces): list of strings
      - [`name`](#terraform-cloud-workspaces): string
      - [`project`](#terraform-cloud-workspaces): string
    - [`hostname`](#terraform-cloud-workspaces): string | `app.terraform.io`

>     - [`token`](#terraform-cloud-token):

Reviewer: ```suggestion
    - [`token`](#terraform-cloud-token): string
```

Modules with experiments enabled generate a warning on every `terraform plan` or `terraform apply` operation. If you want to try experimental features in a shared module, we recommend enabling the experiment only in alpha or beta releases of the module.

Refer to the [Terraform changelog](https://github.com/hashicorp/terraform/blob/main/CHANGELOG.md) for information about experiments and to monitor the release notes about experiment keywords that may be available.

- Data type: List of strings

## Environment variables for the `cloud` block

You can use environment variables to configure one or more `cloud` block attributes. This is helpful when you want to use the same Terraform configuration in different HCP Terraform organizations and projects. Terraform only uses these variables if you do not define  corresponding attributes in your configuration. If you choose to configure the `cloud` block entirely through environment variables, you must still add an empty `cloud` block in your configuration file.

<Warning>

> Remote execution with non-interactive workflows requires auto-approved deployments. Minimize risk of unpredictable infrastructure changes and configuration drift by making sure that no one can change your infrastructure outside of your automated build pipeline.  Refer to [Non-Interactive Workflows](/terraform/cloud-docs/run/cli#non-interactive-workflows) for details.

Author: ```suggestion
You can use environment variables to automate Terraform operations, which has specific security considerations. Refer to [Non-Interactive Workflows](/terraform/cloud-docs/run/cli#non-interactive-workflows) for details.
```
Is something like this what you had in mind?

</Warning>

Use the following environment variables to configure the `cloud` block:

- `TF_CLOUD_ORGANIZATION` - The name of the organization. Terraform reads this variable when `organization` is omitted from the `cloud` block`. If both are specified, the configuration takes precedence.

- `TF_CLOUD_HOSTNAME` - The hostname of a Terraform Enterprise installation. Terraform reads this when `hostname` is omitted from the `cloud` block. If both are specified, the configuration takes precedence.

- `TF_CLOUD_PROJECT` - The name of an HCP Terraform project. Terraform reads this when `workspaces.project` is omitted from the `cloud` block. If both are specified, the cloud block configuration takes precedence.

> - `TF_WORKSPACE` - The name of a single HCP Terraform workspace. Terraform reads this when `workspaces` is omitted from the `cloud` block. HCP Terraform will not create a new workspace from this variable; the workspace must exist in the specified organization. You can set `TF_WORKSPACE` if the `cloud` block uses tags. However, the value of `TF_WORKSPACE` must be included in the set of tags. This variable also selects the workspace in your local environment. Refer to [TF_WORKSPACE](/terraform/cli/config/environment-variables#tf_workspace) for details.

Reviewer 2: Boing back to the comments above about when, and how, HCP-T creates workspaces with the cloud block, reading this I'm unsure if the behavior is the same if you specify the workspace name with the env var or not. I'd expect the behavior to be identical when setting something explicitly vs. using the environment variable, but this makes me think they work slightly differently. If they work the same, we should discuss the ins and outs above where we discuss the feature. If not, I think we need to be explicit in both places.

Author: I agree. Let's follow up with someone after publishing these changes to get clarity.

    workspaces {
      tags = ["app"]
    }
  }
}

```

### Connect to Terraform Enterprise

In the following example, the configuration links the working directory to workspaces in the `example_corp` organization that contain the `app` tag. The `hostname` field is required in the configuration unless you use the `TF_CLOUD_HOSTNAME` environment variable:

```hcl
terraform {
  cloud {
    organization = "example_corp"
    hostname = "my.terraform-enterprise.host"
    workspaces {
      tags = ["app"]
    }
  }
}

```

### Connect to Terraform Enterprise using environment variables

In the following example, local values stored in the `TF_CLOUD_ORGANIZATION` and `TF_CLOUD_HOSTNAME` environment variables automatically populate the configuration. During initialization, the local Terraform CLI connects the working directory to Terraform Enterprise using local values that would otherwise be hard-coded using the `organization` and `hostname` arguments. This allows the configuration to link to ephemeral hostnames and allows teams to  reuse the configuration in different continuous integration pipelines:

```hcl
terraform {
  cloud {
    workspaces {
      tags = ["app"]
    }
  }

> }
> 

Reviewer 3: extra last lines

```suggestion
}
```
